### PR TITLE
Add searching and merging functions

### PR DESCRIPTION
**Pull request by upsj**
To simplify reviewing the ParILUT branch, this PR contains a few cherry-picked algorithms that are used within ParILUT, but are also rather general-purpose. The kernels are based on ideas from Green et al., “GPU merge path: a GPU merging algorithm”

* Searching algorithms that, for a range and a corresponding predicate that is false on the lower part of the range and true on the upper part, identify the first index where the predicate is true. This is the most general form of a binary/k-ary search. There are three implementation variants as mentioned in the paper. (We could also potentially use them in our merge-path SpMV, but that's an idea for another time)

* Merging algorithms that merge two sorted ranges into a single sorted range (via a callback function)

TODO:
- [x] Add references to the implementations
- [x] Remove unnecessary changes

### CODE REVIEW

File: common/components/merging.hpp.inc [hunk 1]
```diff
@@ -0,0 +1,237 @@
...
+(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+******************************<GINKGO LICENSE>*******************************/
+
+namespace detail {
+
+
+/**
+ * @internal
+ * The result from the @ref parallel_merge_step function.
```
Comment (upsj) [inline common/components/merging.hpp.inc]:
```suggestion
 * The result from the @ref group_merge_step function.
```

File: cuda/test/components/searching.cu [hunk 2]
```diff
@@ -0,0 +1,244 @@
...
+
+TEST_F(Searching, EmptySyncBinary)
+{
+    run_test(test_empty_sync_binary_search, 0, config::warp_size);
+}
+
+
+__global__ void test_warp_ary_search(bool *success, int offset, int size)
+{
+    // test binary search on [0, length)
```
Comment (tcojean) [inline cuda/test/components/searching.cu]: length is size I guess, also isn't it from offset rather than 0?

File: cuda/test/components/searching.cu [hunk 3]
```diff
@@ -0,0 +1,244 @@
...
+{
+    for (auto size : sizes) {
+        run_test(test_warp_ary_search, 134, size, size + 1);
+    }
+}
+
+
+__global__ void test_warp_wide_search(bool *success, int offset, int size)
+{
+    // test binary search on [0, length)
```
Comment (tcojean) [inline cuda/test/components/searching.cu]: same

File: cuda/test/components/searching.cu (hunk 2 above)
Comment (tcojean) [reply cuda/test/components/searching.cu]: Maybe these comments are better put in the respective TEST_F, since there is also a discrepancy  with the [0, warp_size) one.

File: common/components/searching.hpp.inc [hunk 4]
```diff
@@ -0,0 +1,230 @@
...
+ * @param p  the predicate to be evaluated on the range - it should not have
+ *           side-effects and map from `IndexType` to `bool`
+ * @returns  the index of `middle`, i.e., the partition point
+ */
+template <typename IndexType, typename Group, typename Predicate>
+__device__ IndexType group_ary_search(IndexType offset, IndexType length,
+                                      Group group, Predicate p)
+{
+    IndexType end = offset + length;
+    // invariant: [offset, offset + length] contains middle
```
Comment (tcojean) [inline common/components/searching.hpp.inc]: I'm unsure of the phrasing here. Technically, it's true this is the pivot space (where a pivot could rightfully exist according to the definition), but "offset+length" is of course out of boundaries of the tested space.  Maybe it would be useful to specifically note somewhere that if the predicate isn't satisfied anywhere, then the pivot has to be `offset + length`.

Comment (upsj) [reply common/components/searching.hpp.inc]: I added a clarifying comment to the function documentations.

File: cuda/components/merging.cuh [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+******************************<GINKGO LICENSE>*******************************/
+
+#ifndef GKO_CUDA_COMPONENTS_MERGING_CUH_
+#define GKO_CUDA_COMPONENTS_MERGING_CUH_
+
+
+#include "cuda/components/intrinsics.cuh"
+#include "cuda/components/searching.cuh"
```
[the remark this answers is not in the document]
Comment (upsj) [reply cuda/components/merging.cuh]: Good point, thanks!

File: hip/components/searching.hip.hpp [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+******************************<GINKGO LICENSE>*******************************/
+
+#ifndef GKO_HIP_COMPONENTS_SEARCHING_CUH_
+#define GKO_HIP_COMPONENTS_SEARCHING_CUH_
+
+
+#include "hip/base/config.hip.hpp"
+#include "hip/components/cooperative_groups.hip.hpp"
```
[the remark this answers is not in the document]
Comment (upsj) [reply hip/components/searching.hip.hpp]: Good point, thanks!

File: cuda/test/components/merging.cu [hunk 7]
```diff
@@ -0,0 +1,287 @@
...
+
+    void init_data()
+    {
+        std::uniform_int_distribution<gko::int32> dist(0, max_size);
+        std::fill_n(data1.get_data(), max_size, 0);
+        std::fill_n(data2.get_data(), max_size, 0);
+        for (auto i = 0; i < max_size; ++i) {
+            // here we also want to test some corner cases
+            // first two runs: zero data1
+            if (rng_run > 0) data1.get_data()[i] = dist(rng);
```
[the remark this answers is not in the document]
Comment (upsj) [reply cuda/test/components/merging.cu]: That's true, I fixed it, thanks!

File: cuda/test/components/merging.cu [hunk 8]
```diff
@@ -0,0 +1,287 @@
...
+    auto warp = tiled_partition<config::warp_size>(this_thread_block());
+    auto i = warp.thread_rank();
+    auto result = group_merge_step<config::warp_size>(a[i], b[i], warp);
+    c[i] = min(result.a_val, result.b_val);
+}
+
+TEST_F(Merging, MergeStep)
+{
+    for (auto i = 0; i < rng_runs; ++i) {
+        init_data();
```
[the remark this answers is not in the document]
Comment (upsj) [reply cuda/test/components/merging.cu]: Good point, I moved it to a parameter.

File: cuda/test/components/searching.cu [hunk 9]
```diff
@@ -0,0 +1,244 @@
...
+class Searching : public ::testing::Test {
+protected:
+    Searching()
+        : ref(gko::ReferenceExecutor::create()),
+          cuda(gko::CudaExecutor::create(0, ref)),
+          result(ref, 1),
+          dresult(cuda),
+          sizes(14203)
+    {
+        std::iota(sizes.begin(), sizes.end(), 0);
```
Comment (tcojean) [inline cuda/test/components/searching.cu]: iota needs `algorithm` include

Comment (upsj) [reply cuda/test/components/searching.cu]: while I think some stdlib implementations include it in `algorithm`, it is supposed to be located in `numeric`

File: common/components/merging.hpp.inc [hunk 10]
```diff
@@ -0,0 +1,237 @@
...
+    /** The element of b being merged in the current thread. */
+    ValueType b_val;
+    /** The index from a that is being merged in the current thread. */
+    int a_idx;
+    /** The index from b that is being merged in the current thread. */
+    int b_idx;
+    /** The number of elements from a that have been merged in total. */
+    int a_advance;
+    /** The number of elements from b that have been merged in total. */
+    int b_advance;
```
[the remark this answers is not in the document]
Comment (upsj) [reply common/components/merging.hpp.inc]: These indices here and in `group_merge_step` are only warp-local (<= config::warp_size), so I wanted to keep the indexing separate. What do you think?

File: cuda/test/components/searching.cu (hunk 9 above)
Comment (tcojean) [reply cuda/test/components/searching.cu]: cppreference says algorithm
https://en.cppreference.com/w/cpp/algorithm/iota

Comment (upsj) [reply cuda/test/components/searching.cu]: > Defined in header <numeric>

Comment (tcojean) [reply cuda/test/components/searching.cu]: my bad, you're right, it's algorithm category but numeric header

File: common/components/merging.hpp.inc (hunk 10 above)
[the remark this answers is not in the document]
Comment (tcojean) [reply common/components/merging.hpp.inc]: Yes, I guess that makes sense. They are also used in a simple way so I guess there is no need to upgrade them to index type when using them in `group_merge`.

File: hip/test/components/merging.hip.cpp [hunk 11]
```diff
@@ -0,0 +1,294 @@
...
+          drefidxs3(hip, 2 * max_size),
+          doutdata(hip, 2 * max_size)
+    {}
+
+    void init_data(int rng_run)
+    {
+        std::uniform_int_distribution<gko::int32> dist(0, max_size);
+        std::fill_n(data1.get_data(), max_size, 0);
+        std::fill_n(data2.get_data(), max_size, 0);
+        for (auto i = 0; i < max_size; ++i) {
```
Comment (thoasm) [inline hip/test/components/merging.hip.cpp]: Please no `auto` in combination with an integer literal.
```suggestion
        for (int i = 0; i < max_size; ++i) {
```

File: cuda/test/components/merging.cu [hunk 12]
```diff
@@ -0,0 +1,285 @@
...
+          drefidxs3(cuda, 2 * max_size),
+          doutdata(cuda, 2 * max_size)
+    {}
+
+    void init_data(int rng_run)
+    {
+        std::uniform_int_distribution<gko::int32> dist(0, max_size);
+        std::fill_n(data1.get_data(), max_size, 0);
+        std::fill_n(data2.get_data(), max_size, 0);
+        for (auto i = 0; i < max_size; ++i) {
```
Comment (thoasm) [inline cuda/test/components/merging.cu]:
```suggestion
        for (int i = 0; i < max_size; ++i) {
```

File: cuda/test/components/merging.cu [hunk 13]
```diff
@@ -0,0 +1,285 @@
...
+{
+    auto warp = tiled_partition<config::warp_size>(this_thread_block());
+    auto i = warp.thread_rank();
+    auto result = group_merge_step<config::warp_size>(a[i], b[i], warp);
+    c[i] = min(result.a_val, result.b_val);
+}
+
+TEST_F(Merging, MergeStep)
+{
+    for (auto i = 0; i < rng_runs; ++i) {
```
Comment (thoasm) [inline cuda/test/components/merging.cu]:
```suggestion
    for (int i = 0; i < rng_runs; ++i) {
```

File: cuda/test/components/merging.cu [hunk 14]
```diff
@@ -0,0 +1,285 @@
...
+    group_merge<config::warp_size>(
+        a, size, b, size, warp,
+        [&](int a_idx, gko::int32 a_val, int b_idx, gko::int32 b_val, int i) {
+            c[i] = min(a_val, b_val);
+        });
+}
+
+TEST_F(Merging, FullMerge)
+{
+    for (auto i = 0; i < rng_runs; ++i) {
```
Comment (thoasm) [inline cuda/test/components/merging.cu]:
```suggestion
    for (int i = 0; i < rng_runs; ++i) {
```

File: cuda/test/components/merging.cu [hunk 15]
```diff
@@ -0,0 +1,285 @@
...
+    sequential_merge(
+        a, size, b, size,
+        [&](int a_idx, gko::int32 a_val, int b_idx, gko::int32 b_val, int i) {
+            c[i] = min(a_val, b_val);
+        });
+}
+
+TEST_F(Merging, SequentialFullMerge)
+{
+    for (auto i = 0; i < rng_runs; ++i) {
```
Comment (thoasm) [inline cuda/test/components/merging.cu]:
```suggestion
    for (int i = 0; i < rng_runs; ++i) {
```

File: cuda/test/components/merging.cu [hunk 16]
```diff
@@ -0,0 +1,285 @@
...
+            aidxs[i] = a_idx;
+            bidxs[i] = b_idx;
+            cidxs[i] = i;
+            c[i] = min(a_val, b_val);
+        });
+}
+
+TEST_F(Merging, FullMergeIdxs)
+{
+    for (auto i = 0; i < rng_runs; ++i) {
```
Comment (thoasm) [inline cuda/test/components/merging.cu]:
```suggestion
    for (int i = 0; i < rng_runs; ++i) {
```

File: hip/test/components/merging.hip.cpp [hunk 17]
```diff
@@ -0,0 +1,294 @@
...
+{
+    auto warp = tiled_partition<config::warp_size>(this_thread_block());
+    auto i = warp.thread_rank();
+    auto result = group_merge_step<config::warp_size>(a[i], b[i], warp);
+    c[i] = min(result.a_val, result.b_val);
+}
+
+TEST_F(Merging, MergeStep)
+{
+    for (auto i = 0; i < rng_runs; ++i) {
```
Comment (thoasm) [inline hip/test/components/merging.hip.cpp]:
```suggestion
    for (int i = 0; i < rng_runs; ++i) {
```

File: hip/test/components/merging.hip.cpp [hunk 18]
```diff
@@ -0,0 +1,294 @@
...
+    group_merge<config::warp_size>(
+        a, size, b, size, warp,
+        [&](int a_idx, gko::int32 a_val, int b_idx, gko::int32 b_val, int i) {
+            c[i] = min(a_val, b_val);
+        });
+}
+
+TEST_F(Merging, FullMerge)
+{
+    for (auto i = 0; i < rng_runs; ++i) {
```
Comment (thoasm) [inline hip/test/components/merging.hip.cpp]:
```suggestion
    for (int i = 0; i < rng_runs; ++i) {
```

File: hip/test/components/merging.hip.cpp [hunk 19]
```diff
@@ -0,0 +1,294 @@
...
+    sequential_merge(
+        a, size, b, size,
+        [&](int a_idx, gko::int32 a_val, int b_idx, gko::int32 b_val, int i) {
+            c[i] = min(a_val, b_val);
+        });
+}
+
+TEST_F(Merging, SequentialFullMerge)
+{
+    for (auto i = 0; i < rng_runs; ++i) {
```
Comment (thoasm) [inline hip/test/components/merging.hip.cpp]:
```suggestion
    for (int i = 0; i < rng_runs; ++i) {
```

File: hip/test/components/merging.hip.cpp [hunk 20]
```diff
@@ -0,0 +1,294 @@
...
+            aidxs[i] = a_idx;
+            bidxs[i] = b_idx;
+            cidxs[i] = i;
+            c[i] = min(a_val, b_val);
+        });
+}
+
+TEST_F(Merging, FullMergeIdxs)
+{
+    for (auto i = 0; i < rng_runs; ++i) {
```
Comment (thoasm) [inline hip/test/components/merging.hip.cpp]:
```suggestion
    for (int i = 0; i < rng_runs; ++i) {
```

File: hip/test/components/searching.hip.cpp [hunk 21]
```diff
@@ -0,0 +1,249 @@
...
+        auto success = *result.get_const_data();
+
+        ASSERT_TRUE(success);
+    }
+
+    std::shared_ptr<gko::ReferenceExecutor> ref;
+    std::shared_ptr<gko::HipExecutor> hip;
+    gko::Array<bool> result;
+    gko::Array<bool> dresult;
+    std::vector<int> sizes;
```
Comment (thoasm) [inline hip/test/components/searching.hip.cpp]: You need to include `vector` for this one.

File: hip/test/components/merging.hip.cpp [hunk 22]
```diff
@@ -0,0 +1,294 @@
...
+SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+******************************<GINKGO LICENSE>*******************************/
+
+// TODO remove when the HIP includes are fixed
+#include <hip/hip_runtime.h>
```
Comment (thoasm) [inline hip/test/components/merging.hip.cpp]: Is this required to be included first, or could it be together with other third party includes?

File: common/components/searching.hpp.inc [hunk 23]
```diff
@@ -0,0 +1,235 @@
...
+ * @internal
+ * Generic search that finds the first index where a predicate is true.
+ * It assumes that the predicate partitions the range [offset, offset + length)
+ * into two subranges [offset, middle), [middle, offset + length) such that
+ * the predicate is `false` for all elements in the first range and `true` for
+ * all elements in the second range. `middle` is called the partition point.
+ * If the predicate is `false` everywhere, `middle` equals `offset + length`.
+ *
+ * It executes `log2(length) / log2(group.size())` calls to `p` that effectively
+ * follow a random-access pattern.
```
[the remark this answers is not in the document]
Comment (upsj) [reply common/components/searching.hpp.inc]: I would rather leave that as it is (calls [...] that [...] follow)

File: hip/test/components/merging.hip.cpp (hunk 22 above)
Comment (upsj) [reply hip/test/components/merging.hip.cpp]: Yes, some HIP files are not self-contained, so they need the include from hip_runtime.h

File: common/components/merging.hpp.inc [hunk 24]
```diff
@@ -0,0 +1,237 @@
...
+        return a_remote >= b_remote || x >= i;
+    });
+
+    auto a_idx = minx;
+    auto b_idx = max(i - minx, 0);
+    auto a_val = group.shfl(a, a_idx);
+    auto b_val = group.shfl(b, b_idx);
+    auto cmp = a_val < b_val;
+    auto a_advance = int(popcnt(group.ballot(cmp)));
+    auto b_advance = int(group.size()) - a_advance;
```
[the remark this answers is not in the document]
Comment (upsj) [reply common/components/merging.hpp.inc]: all involved values are <= 64, so I would not expect any issues here. Also, `group.size()` returns an unsigned value, so this conversion would fail. I could remove the cast from `popcnt` since it returns an `int` anyways, but would like to leave the rest as it is.

File: cuda/test/components/merging.cu [hunk 25]
```diff
@@ -0,0 +1,285 @@
...
+(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+******************************<GINKGO LICENSE>*******************************/
+
+#include "cuda/components/merging.cuh"
+
+
+#include <algorithm>
+#include <memory>
+#include <random>
```
Comment (upsj) [inline cuda/test/components/merging.cu]:
```suggestion
#include <random>
#include <vector>
```

File: cuda/test/components/searching.cu [hunk 26]
```diff
@@ -0,0 +1,244 @@
...
+THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+******************************<GINKGO LICENSE>*******************************/
+
+#include "cuda/components/searching.cuh"
+
+
+#include <memory>
+#include <numeric>
```
Comment (upsj) [inline cuda/test/components/searching.cu]:
```suggestion
#include <numeric>
#include <vector>
```

File: hip/test/components/merging.hip.cpp [hunk 27]
```diff
@@ -0,0 +1,294 @@
...
+
+
+#include "hip/components/merging.hip.hpp"
+
+
+#include <algorithm>
+#include <memory>
+#include <random>
```
Comment (upsj) [inline hip/test/components/merging.hip.cpp]:
```suggestion
#include <random>
#include <vector>
```

File: hip/test/components/searching.hip.cpp [hunk 28]
```diff
@@ -0,0 +1,249 @@
...
+
+// TODO remove when the HIP includes are fixed
+#include <hip/hip_runtime.h>
+
+
+#include "hip/components/searching.hip.hpp"
+
+
+#include <memory>
+#include <numeric>
```
Comment (upsj) [inline hip/test/components/searching.hip.cpp]:
```suggestion
#include <numeric>
#include <vector>
```

File: hip/test/components/searching.hip.cpp (hunk 21 above)
Comment (upsj) [reply hip/test/components/searching.hip.cpp]: Good point, thanks!

File: common/components/merging.hpp.inc (hunk 24 above)
[the remark this answers is not in the document]
Comment (thoasm) [reply common/components/merging.hpp.inc]: If you really want to keep the current status, I will not stop you from doing so.
I simply dislike any kind of C-style casts, especially if it is easy to replace it with a C++ cast.

File: common/components/merging.hpp.inc [hunk 29]
```diff
@@ -0,0 +1,237 @@
...
+        auto b_remote = group.shfl(b, max(i - x - 1, 0));
+        return a_remote >= b_remote || x >= i;
+    });
+
+    auto a_idx = minx;
+    auto b_idx = max(i - minx, 0);
+    auto a_val = group.shfl(a, a_idx);
+    auto b_val = group.shfl(b, b_idx);
+    auto cmp = a_val < b_val;
+    auto a_advance = int(popcnt(group.ballot(cmp)));
```
Comment (upsj) [inline common/components/merging.hpp.inc]:
```suggestion
    auto a_advance = popcnt(group.ballot(cmp));
```

Comment (thoasm) [reply common/components/merging.hpp.inc]: That is fine with me.
Just out of curiosity: Why are you proposing changes to your own PR? To keep taps on it?

Comment (upsj) [reply common/components/merging.hpp.inc]: Yes, I wanted to add all proposed small changes to a single commit instead of having to manually apply them on my system

File: common/components/merging.hpp.inc (hunk 24 above)
[the remark this answers is not in the document]
Comment (upsj) [reply common/components/merging.hpp.inc]: For casting integer types, `static_cast<...>` looks unnecessarily verbose to me. In case you are not aware, for non-cv qualified integer types, a functional-style cast effectively is a `static_cast`, since [all other options](https://en.cppreference.com/w/cpp/language/explicit_cast) (const_cast, reinterpret_cast) do not apply there

**Comment by yhmtsai** [comment]
non self-contained hip file is `hip/device_functions.h` which needs to include `hip/hip_runtime.h` first.
We only use `hip/device_functions.h` in the `config.hip.hpp` to get warpsize.
For temporary fix, put the `hip/hip_runtime.h` into `config.hip.hpp` before `hip/device_functions.h`.
It should fix the problem, and I also did it in PR #447 sorting and config changes (https://github.com/ginkgo-project/ginkgo/pull/447/files/55a80da0a4247401db21d5ae4932930a3d01b176)
Could you also try it here?

File: common/components/merging.hpp.inc (hunk 24 above)
[the remark this answers is not in the document]
Comment (thoasm) [reply common/components/merging.hpp.inc]: Yes, I know that. However, I still prefer the C++ casts for 2 reasons:
1. For consistency, so it is unlikely I ever make the mistake to do a functional cast with a template parameter or an unknown type
2. I personally like the verbosity of these casts, since it is extremely easy to spot them (especially during debugging). Everyone who is reading `static_cast` knows instantly that a cast is happening, while it is easy to over-read `int()`.

[the remark this answers is not in the document]
Comment (upsj) [reply common/components/merging.hpp.inc]: I see, I need to think about the implications a bit. But would you consider it okay to leave this cast in?

[the remark this answers is not in the document]
Comment (thoasm) [reply common/components/merging.hpp.inc]: Yes, you can leave it in, it was just a sugesetion and I don't consider it absolutely mandatory to use C++ style casts (**in this context**).
As you said, there is pretty much next to no way that this cast can cause any issues, so you can leave it in.

File: common/components/intrinsics.hpp.inc [hunk 30]
```diff
@@ -34,21 +34,21 @@ OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
...
+inline __device__ int popcnt(uint64 mask) { return __popcll(mask); }
 
 
 /**
  * @internal
  * Returns the (1-based!) index of the first set bit in the given mask,
  * starting from the least significant bit.
  */
-__device__ int ffs(uint32 mask) { return __ffs(mask); }
+inline __device__ int ffs(uint32 mask) { return __ffs(mask); }
```
Comment (thoasm) [inline common/components/intrinsics.hpp.inc]: Usually, we use `__forceinline__` for inlining in GPU kernels (at least we always did it in CUDA and we currently use it in pretty much all `common` kernels like in `atomic` or `cooperative_groups`).

File: common/components/intrinsics.hpp.inc [hunk 31]
```diff
@@ -34,21 +34,21 @@ OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
  * @internal
  * Returns the number of set bits in the given mask.
  */
-__device__ int popcnt(uint32 mask) { return __popc(mask); }
+inline __device__ int popcnt(uint32 mask) { return __popc(mask); }
```
Comment (upsj) [inline common/components/intrinsics.hpp.inc]:
```suggestion
__forceinline__ __device__ int popcnt(uint32 mask) { return __popc(mask); }
```

File: common/components/intrinsics.hpp.inc [hunk 32]
```diff
@@ -34,21 +34,21 @@ OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
...
 
 /** @copydoc popcnt */
-__device__ int popcnt(uint64 mask) { return __popcll(mask); }
+inline __device__ int popcnt(uint64 mask) { return __popcll(mask); }
```
Comment (upsj) [inline common/components/intrinsics.hpp.inc]:
```suggestion
__forceinline__ __device__ int popcnt(uint64 mask) { return __popcll(mask); }
```

File: common/components/intrinsics.hpp.inc [hunk 33]
```diff
@@ -34,21 +34,21 @@ OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
...
  * @internal
  * Returns the (1-based!) index of the first set bit in the given mask,
  * starting from the least significant bit.
  */
-__device__ int ffs(uint32 mask) { return __ffs(mask); }
+inline __device__ int ffs(uint32 mask) { return __ffs(mask); }
 
 /** @copydoc ffs */
-__device__ int ffs(uint64 mask)
+inline __device__ int ffs(uint64 mask)
```
Comment (upsj) [inline common/components/intrinsics.hpp.inc]:
```suggestion
__forceinline__ __device__ int ffs(uint64 mask)
```

File: common/components/intrinsics.hpp.inc [hunk 34]
```diff
@@ -60,7 +60,7 @@ __device__ int ffs(uint64 mask)
  * Returns the number of zero bits before the first set bit in the given mask,
  * starting from the most significant bit.
  */
-__device__ int clz(uint32 mask) { return __clz(mask); }
+inline __device__ int clz(uint32 mask) { return __clz(mask); }
```
Comment (upsj) [inline common/components/intrinsics.hpp.inc]:
```suggestion
__forceinline__ __device__ int clz(uint32 mask) { return __clz(mask); }
```

File: common/components/intrinsics.hpp.inc [hunk 35]
```diff
@@ -60,7 +60,7 @@ __device__ int ffs(uint64 mask)
...
 
 /** @copydoc clz */
-__device__ int clz(uint64 mask) { return __clzll(mask); }
+inline __device__ int clz(uint64 mask) { return __clzll(mask); }
```
Comment (upsj) [inline common/components/intrinsics.hpp.inc]:
```suggestion
__forceinline__ __device__ int clz(uint64 mask) { return __clzll(mask); }
```
